Improve status handling

stlaz opened the pull request:
This will block the installation until we report availability

enj commented on manifests/cluster-authentication-operator_06_clusteroperator.yaml:
```diff
@@ -0,0 +1,5 @@
+kind: ClusterOperator
+metadata:
+  name: openshift-authentication
+spec: {}
```
omit all together ?

stlaz replied:
I shamelessly copied this from kube-apiserver-operator, let's see what happens when we remove it.

nee1esh commented:
/retest

nee1esh commented:
Looks like there are some syntax/build errors in pkg/operator2/operatorclient.go

stlaz commented:
CI is probably unable to apply the changes due to the required rebase (the unexpected == and << come from merge conflicts). We need to get status reporting right first anyway before merging this PR.

enj commented:
The CO code we have is buggy and incomplete because #36 was merged before it was ready.  This needs to wait until those issues are resolved.

stlaz commented:
I pushed just to create backup, the order of commits is not ok and the operator will be in a beautiful update cycle as we're updating the operator's config by the status, which in turn tries to update the osin deployment, which leads to updating the status.

I'll try to sort the cyclic issue by using generations where possible tomorrow.

stlaz commented:
The cycles are resolved, it might be better to join generation + resource kind to make sure that the value is more unique, but it seems that the installer properly waits for us to be ready \o/

stlaz commented:
@openshift/sig-auth Probably good/suitable for review. I've focused mostly on reporting availability when I believe we're sure to be serving authn requests properly (deployment is ready, the route gives 200s and the well-known endpoint advertises expected values).

The Failing state is still quite underdeveloped.

Also, see if you're fine with how redeployments are done now in relation to objects that include generation. This needed to change as "resourceVersion" changed with any status object of such object, whereas "generation" does not do that.

Note that this also moves our operator back to the default namespace as required per discussion on the team list.

stlaz commented:
/retest

enj commented on pkg/operator2/operator.go:
```diff
@@ -311,3 +431,34 @@ func getPrefixFilter() controller.Filter {
 	}
 }
+
+func newHTTPSClient() (*http.Client, error) {
```
Actually you do need to keep rebuilding since this will talk to a route.  As we discussed on slack, the TransportFor rest config logic is probably what you want.

stlaz replied:
I keep the Transport and rebuild the client, not exactly sure whether I need to rebuild the Transport too. It seems to do the trick as is, though.

stlaz commented:
Fixed the comments. I thought it might be good to see what happens during the progressing status so I added setting reason to it in the waiting stages.

enj commented on pkg/operator2/operator.go:
```diff
@@ -283,16 +317,186 @@ func (c *authOperator) handleSync(operatorConfig *operatorv1.Authentication) err
 	glog.V(4).Infof("current deployment: %#v", deployment)
 
+	if changed {
+		if err = c.setProgressingStatus(operatorConfig, "DeploymentChanged", ""); err != nil {
```
We should hit this branch whenever `changed || replicas != expected`

enj commented on pkg/operator2/status.go:
```diff
@@ -33,6 +33,28 @@ func (c *authOperator) setFailingStatus(operatorConfig *operatorv1.Authenticatio
 	return err
 }
 
+func (c *authOperator) setProgressingStatus(operatorConfig *operatorv1.Authentication, reason, message string) error {
```
I am wondering if we call this too much - i.e. if we should build up a series of conditions and apply once?

enj replied:
Actually it should not matter because the helper functions only update when needed.

enj commented on pkg/operator2/operator.go:
```diff
@@ -283,16 +317,186 @@ func (c *authOperator) handleSync(operatorConfig *operatorv1.Authentication) err
 }
 
+func (c *authOperator) WaitUntilReady(operatorConfig *operatorv1.Authentication, route *routev1.Route, deploymentVersionHash string) error {
+	return wait.Poll(1, 5*time.Minute, func() (bool, error) {
```
Max time should be around 2 mins.

enj commented on manifests/cluster-authentication-operator_06_clusteroperator.yaml:
```diff
@@ -0,0 +1,4 @@
+apiVersion: config.openshift.io/v1
+kind: ClusterOperator
+metadata:
+  name: authentication
```
Ex version: https://github.com/openshift/cluster-image-registry-operator/pull/225

stlaz replied:
Currently I'm focusing only on getting the conditions right.

We haven't worked with generations very much up until now, and it's still far away from being optimal.

I'll see about getting version in here.

stlaz commented on pkg/operator2/operator.go:
```diff
@@ -206,7 +239,7 @@ func (c *authOperator) handleSync(operatorConfig *operatorv1.Authentication) err
 		return err
 	}
-	resourceVersions = append(resourceVersions, authConfig.GetResourceVersion())
+	generationSum += authConfig.Generation
```
The thing is that a hash of the same number of a different resource will always be the same so it does not matter whether you sum it or keep them separate.

We could however do separate `generation + kind` strings to distinguish this properly.

stlaz replied:
Ah, never mind, since we do `rvs := strings.Join(resourceVersions, ",")` it's already order-dependent. It would have been anyway.

stlaz commented:
Addressed comments, should still add version.

stlaz commented:
Did some substantial changes to the operator status updating logic. We now also track the generation of our deployment.

Still need to add version.

stlaz commented:
Added version, turned out to be very easy.

stlaz commented:
Based on [e2e-aws/results/clusteroperators.json/](https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_cluster-authentication-operator/77/pull-ci-openshift-cluster-authentication-operator-master-e2e-aws/416/artifacts/e2e-aws/clusteroperators.json), the version seems to work.

stlaz commented:
/retest

nee1esh commented:
/retest

stlaz commented:
Hm, I see that both the operator and its pods ran into cert errors. Weird.

enj commented on manifests/cluster-authentication-operator_03_version-mapping.yaml:
```diff
@@ -0,0 +1,8 @@
+kind: ConfigMap
+metadata:
+  namespace: openshift-authentication-operator
+  name: version-mapping
```
Do we need this?  I think it is only used in the `VersionForOperand` code.

stlaz replied:
I couldn't find `version-mapping` code, was looking at bad places (CVO repo) :man_facepalming: I'll use this instead of the `IMAGE_VERSION` env var in the payload deployment.

enj commented on pkg/operator2/operator.go:
```diff
@@ -379,10 +378,16 @@ func (c *authOperator) CheckReady(
 	}
 	if !deploymentReady {
-		c.setProgressingStatus(operatorConfig, "DeployementNotReady", deploymentMsg)
+		c.setProgressingStatus(operatorConfig, "DeploymentNotReady", deploymentMsg)
```
Might want to be specific on which deployment.

enj commented on pkg/operator2/configmap.go:
```diff
@@ -37,6 +38,20 @@ const stubMetadata = `
+	metadataJSON := getMetadata(route)
+	err := json.Unmarshal([]byte(metadataJSON), &ret)
+	if err != nil {
+		panic(err)
```
comment stating this panic should never happen unless our static metadata is broken

enj commented on pkg/operator2/operator.go:
```diff
@@ -180,34 +212,30 @@ func (c *authOperator) Sync(obj metav1.Object) error {
-	operatorConfigCopy := operatorConfig.DeepCopy()
-
-	syncErr := c.handleSync(operatorConfigCopy)
+	syncErr := c.handleSync(operatorConfig)
```
Keep the copy.  We want to always know that this is safe to mutate.

stlaz commented on pkg/operator2/operator.go:
```diff
@@ -180,34 +212,30 @@ func (c *authOperator) Sync(obj metav1.Object) error {
-			return err
-		}
-	}
+	v1helpers.UpdateStatus(c.authOperatorConfigClient, func(status *operatorv1.OperatorStatus) error {
```
There was a bug caused by rebase, I forgot I should be setting the available status somewhere else. This is for all the cases.

stlaz commented on pkg/operator2/operator.go:
```diff
@@ -180,34 +212,30 @@ func (c *authOperator) Sync(obj metav1.Object) error {
-		}
-	}
+	v1helpers.UpdateStatus(c.authOperatorConfigClient, func(status *operatorv1.OperatorStatus) error {
+		status = operatorConfig.Status.OperatorStatus.DeepCopy()
```
Right, this is but a beautiful NOP

stlaz commented:
Addressed all comments, submitting to e2e to check that the new version for operand through version-mapping works.

enj commented on pkg/operator2/operator.go:
```diff
@@ -9,10 +9,11 @@ import (
 	"reflect"
 	"strings"
 
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
```
move down

enj commented:
/lgtm
